lib/maths/COutliers.cc

```diff
@@ -9,87 +9,514 @@
+    };
+    return {sampleSize, onSample, rng};
+}
+
+template<typename POINT>
+void CEnsemble<POINT>::CScorer::add(const TMeanVarAccumulatorVec& logScoreMoments,
+                                    const TDoubleVec& scores) {
+
+    // The basic idea is to map score percentile to a probability that the data
+    // point is an outlier. We use a monotonic function for this which is flat
+    // near the median, doesn't become too small for low percentiles and is
+    // relatively sensitive for very high percentiles. Even then we don't become
+    // too confident based on a single score. This could be explicitly calibrated
+    // using the slope of the ROC curve, but one would need a very good collection
+    // of benchmark sets to make this effective a priori. We also account for the
+    // fact that the scores will be somewhat correlated.
+
+    auto pOutlier = [](double cdfComplement) {
+        static const double LOG_KNOTS[]{
+            CTools::fastLog(1e-5), CTools::fastLog(1e-3), CTools::fastLog(0.01),
+            CTools::fastLog(0.1),  CTools::fastLog(0.2),  CTools::fastLog(0.4),
+            CTools::fastLog(0.5),  CTools::fastLog(0.6),  CTools::fastLog(0.8),
+            CTools::fastLog(1.0)};
+        static const double KNOT_PROBABILITIES[]{0.98, 0.87, 0.76, 0.65, 0.6,
+                                                 0.5,  0.5,  0.5,  0.3,  0.3};
+
+        double logCdfComplement{CTools::fastLog(std::max(cdfComplement, 1e-5))};
+        auto k = std::upper_bound(std::begin(LOG_KNOTS), std::end(LOG_KNOTS), logCdfComplement);
+        return CTools::linearlyInterpolate(
+            *(k - 1), *k, KNOT_PROBABILITIES[k - std::begin(LOG_KNOTS) - 1],
+            KNOT_PROBABILITIES[k - std::begin(LOG_KNOTS)], logCdfComplement);
+    };
+
+    static const double EXPECTED_PROBABILITY{[pOutlier] {
+        double result{0.0};
+        for (double x = 0.0; x < 0.99; x += 0.1) {
+            double interval;
+            CIntegration::gaussLegendre<CIntegration::OrderTwo>(
+                [pOutlier](double x_, double& r) {
+                    r = pOutlier(x_);
+                    return true;
+                },
+                x, x + 0.1, interval);
+            result += interval;
+        }
+        return result;
+    }()};
+
+    double logLikelihoodOutlierGivenScores{0.0};
+    double logLikelihoodInlierGivenScores{0.0};
+
+    for (std::size_t i = 0; i < scores.size(); ++i) {
+        double location{CBasicStatistics::mean(logScoreMoments[i])};
+        double scale{std::sqrt(CBasicStatistics::variance(logScoreMoments[i]))};
+        double cdfComplement{0.5};
+        if (scale > 0.0) {
+            try {
+                boost::math::lognormal lognormal{location, scale};
+                cdfComplement = CTools::safeCdfComplement(lognormal, shift(scores[i]));
+            } catch (const std::exception& e) {
+                LOG_ERROR(<< "Failed to normalise scores: " << e.what());
```

Comment:
I thought about this a bit. In this case, the default value assigned P(outlier | score) = 0.5 = P(inlier | score) so this model's result for this point is effectively ignored, i.e. the final conditional probability will be unchanged. I think this is perhaps a good enough way of handling this situation, i.e. useful things could still come out of the analysis and failing completely here seemed a bit extreme. What do you think?

Reply:
I am just careful, users perceive such error message as big problem, maybe we lower the level?

It's a tough decision: Technically `error` is exactly what it is, but as said and as you explained logically it can be ignored.

Reply:
I think on balance I'm happy for this to be a warn given we have a semi effective treatment and my main concern is that it isn't invisible to us in internal QA'ing. This is probably a good case to instrument with a program counter as well and we might want to issue an error if the number became high. I'll leave that change for another pull request.